### ISSUE TITLE
rescore_query not working?  (7.6.2 and 7.9.2)

### ISSUE THREAD

**Issue by ejackson-eb** [alexklibisz/elastiknn#182]
Maybe I'm doing something wrong, but I seem to be getting back all zero scores when using elastiknn_nearest_neighbors in a rescore_query as suggested in the documentation.  Query:

`{'_source': ['id'], 'size': 20, 'query': {'function_score': {'query': {'bool': {'filter': [{'term': {'online_event': 'true'}}], 'must': {'match_all': {}}}}}}, 'rescore': {'window_size': 10, 'query': {'rescore_query': {'elastiknn_nearest_neighbors': {'field': 'esknn_embedding', 'vec': {'values': [...]}, 'model': 'lsh', 'similarity': 'angular', 'candidates': 50}}, 'query_weight': 0, 'rescore_query_weight': 1}}}`

**Comment by alexklibisz** [comment]
Here's the query re-formatted for easier reading:
```
{
  "_source": [
    "id"
  ],
  "size": 20,
  "query": {
    "function_score": {
      "query": {
        "bool": {
          "filter": [
            {
              "term": {
                "online_event": "true"
              }
            }
          ],
          "must": {
            "match_all": {}
          }
        }
      }
    }
  },
  "rescore": {
    "window_size": 10,
    "query": {
      "rescore_query": {
        "elastiknn_nearest_neighbors": {
          "field": "esknn_embedding",
          "vec": {
            "values": [...]
          },
          "model": "lsh",
          "similarity": "angular",
          "candidates": 50
        }
      },
      "query_weight": 0,
      "rescore_query_weight": 1
    }
  }
}
```

**Comment by alexklibisz** [comment]
@ejackson-eb Just to make sure, you are trying to find all docs which have `online_event = true` and then rescore their embeddings? I think there might be a simpler way to express that, without using a function_score. Can you try the following:

```
{
  "_source": [
    "id"
  ],
  "size": 20,
  "query": {
    "term": {
      "online_event": true 
    }
  },
  "rescore": {
    "window_size": 10,
    "query": {
      "rescore_query": {
        "elastiknn_nearest_neighbors": {
          "field": "esknn_embedding",
          "vec": {
            "values": [...]
          },
          "model": "lsh",
          "similarity": "angular",
          "candidates": 50
        }
      },
      "query_weight": 0,
      "rescore_query_weight": 1
    }
  }
}
```

**Comment by ejackson-eb** [comment]
I see the same problem with your query.

**Comment by alexklibisz** [comment]
@ejackson-eb I think I know what's happening. Since this is pretty closely connected to #183 , I'll address some quotes from there.

> 1) The simplest thing is to combine a filter and an "elastiknn_nearest_neighbors" clause in the main body of the query.  For example, something like:
```
     "bool": {
       "filters": [some filters]
      "must": [{"elastiknn_nearest_neighbors": ...}]
     }
```

> What I've noticed is that the performance doesn't really depend on how many documents match the filter.  It looks like Elasticsearch is evaluating elasticknn_nearest_neighbors on all the documents, regardless of whether they satisfy the filter.  That's an inference based on all the timings being the same.

Your observation here is correct and it also tripped me up at first. The `must` clause gets evaluated just like it would without any filters. I think it's a pretty confusing query API. This was actually how I had originally recommended to do pre-filtering, and then I looked under the covers a bit and figured out rescore_query is actually what we need.

---

> Maybe I'm doing something wrong, but I seem to be getting back all zero scores when using elastiknn_nearest_neighbors in a rescore_query as suggested in the documentation.

I think what's happening here is this: you are using a filter followed by an _approximate_ query. Your filter will match at most `window_size` docs and pass those through to the approximate query. Then the approximate query will find at most `candidates` matching vectors. However, it can actually find fewer than that number of candidates depending on your model settings. Specifically, if your model settings cause high precision and low recall, then you're likely to find fewer than the desired number of candidates. That should explain the zeros in your scores. In other words, the model settings cause you to find 0 candidates among the vectors that matched your filter, so none of the scores from the filter query get adjusted.

So I'd recommend the following: 
- Make sure that `candidates >= window_size >= size`. Ideally `candidates = window_size` and `window_size` is much larger than `size`, e.g. 10k and 10.
- Switch to using an exact query. These are highly optimized and should comfortably return in < 500ms for 10k docs. For reference, exact queries on a corpus of 60k 784 dimensional vectors return in about 250ms in the benchmarks I've done.
- If the exact query is too slow, switch back to an approximate query and adjust your model parameters. Most of the models have directions like `Generally, increasing this value increases recall.` or `Generally, increasing this value increases precision.` Expect a bit of trial and error. 

---

> b) you can only pass up to 10,000 documents from the main query to the rescore query (that's the max "window_size").  Sometimes our filters are broad and match more than 10,000 documents.

Unfortunately, if that's a hard-coded limit in Elasticsearch there's not much we can do about it. Switching to painless scripting is possibly an option, but not something that will happen quickly. However, it wasn't obvious to me from the docs that this is a hard limit enforced by rescore query itself. Maybe what you're actually hitting is the overall limit on the `size` parameter? If so, it sounds like there's a way to adjust that: https://discuss.elastic.co/t/how-to-increase-the-default-size-limit-from-10000-to-1000000-in-elasticsearch/208807

**Comment by ejackson-eb** [comment]
When you set the window size over 10,000 you get this:

`org.elasticsearch.ElasticsearchException$1: Rescore window [10001] is too large. It must be less than [10000]. This prevents allocating massive heaps for storing the results to be rescored. This limit can be set by changing the [index.max_rescore_window] index level setting.`

I guess it does say I can change the limit.

Anyhow, I think I can live with this.

**Comment by alexklibisz** [comment]
Sounds good. I opened a PR which adds logger warnings when approximate queries find fewer than the requested number of approximate candidates.

I'll go ahead and get that merged/released for 7.6.2.

Feel free to re-open if this is still an issue.
